source/load-balancers/load-balancers.rst

```diff
@@ -0,0 +1,531 @@
+but not limited to mongos, Atlas Proxy, Elastic Front-End, and Data Lake. All services would need
+to be modified to understand the PROXY protocol (although Data Lake already does), or add
+configuration options that are read during start-up. From there they would need to alter their
+hello responses in order to tell the driver they were behind a load balancer.
+
+From a driver's view there is a change in behaviour in that once in load balanced mode, the driver
+can never transition out of that mode, unlike other states in the SDAM specification.
+
+Why does this specification instruct drivers to not check connections back into the connection pool in some circumstances?
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In the case of a load balancer fronting multiple services, it is possible that a connection to th
+e load balancer could result in a connection behind the load balancer to a different service. In
+order to guarantee these operations execute on the same service they need to be executed on the
+same socket - not checking a connection back into the pool for the entire operation guarantees this.
+
+What reason has a client side connection reaper for idle cursors not been put into this specification?
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+It was discussed as a potential solution for maxed out connection pools that the drivers could
+potentially behave similar to the server and close long running cursors after a specified time
+period and return their connections to the pool. Due to the high complexity of that solution
+it was determined that better error messaging when the connection pool was maxed out would
+suffice in order for users to easily debug when the pool ran out of connections and fix their
+applications or adjust their pool options accordingly.
+
+Why are we requiring mongos servers to add a new serviceId field in hello responses rather than reusing the existing topologyVersion.processId?
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This option was previously discussed, but we opted to add a new :code:`hello` response field in
+order to not mix intentions.
+
+Why does this specification not address load balancer restarts or maintenance?
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+It was decided that the core problem that was needed to be solved was to introduce durable/resumable
+cursors to some services. The related ticket is `DRIVERS-1549 <https://jira.mongodb.org/browse/DRIVERS-1549>`__.
+
+
+Design Rationales
+-----------------
+
+This design is based on the simplest approach to implementation in both the drivers and all services,
+when taking into consideration that it is NOT a requirement that applications connected to any service
```

Comment:
I'm not sure I understand this design rationale.  Can you say more clearly what it is you mean?

The rationale that I recall is that we wanted people to opt in to load balancing mode because it may require changes in other settings, in particular max pool size.  For the same reason, we didn't just want to start pinning connections/sessions in all modes because of connection pool pressure.

Reply:
I've reworded this sections now to be more clear.

Reply:
I still don't get it.  Let's discuss it in person to figure out what we really mean here.

Reply:
Thought about this a bit while out on evening stroll.  I think these are the points we want to make:

Q.  Why do we pin connections to open transactions and cursors

A Because that's the only way to ensure that we can send subsequent messages with same txn/cursor id to the same mongos server that initiated the txn/cursor.

Q. Why do applications have to opt-in to this mode?

A. Load balancing and pinning creates new stresses on connection pools.  We go from a total number of connections to the cluster of num-mongos-servers * max-pool-size to simply max-pool-size.  And in addition, connections get pinned to open cursors and transactions, further straining resource availability.  So in addition to opting in applications my need to bump up the max pool size.

Reply:
IMO the first question (why do we pin) is already answered in the spec's [Connection Pinning](https://github.com/mongodb/specifications/blob/79c138069b47612b610cbb9cf7a972408d0069ea/source/load-balancers/load-balancers.rst#connection-pinning) section. For the second question, I agree that this should be made clearer in the "Design Rationale" section. Suggested wording for this section:

> Other versions of this design proposed a scheme in which the application does not have to opt-in to load balanced mode. Instead, the server would send a special field in ``hello`` and legacy ``hello`` command responses to indicate that it was running behind a load balancer and the driver would change its behavior accordingly. We opted to take an approach that required code changes instead because load balancing changes driver behavior in ways that could cause unexpected application errors, so it made sense to have applications consciously opt-in to this mode.  For example, connection pinning creates new stresses on connection pools because we go from a total of ``numMongosServers * maxPoolSize`` connections to simply ``maxPoolSize``. Furthermore, connections get pinned to open cursors and transactions, further straining resource availability. Due to this change, applications may also need to increase the configured ``maxPoolSize`` when opting into this mode.

Reply:
True about the connection pinning rationale.  It's a little strange to have a rationale in the main portion of the spec.  Maybe it should move here?  Either way is ok with me though.

Reply:
I suggested adding it in the spec itself because it felt weird to say "pin connections in these cases" without at least giving a basic explanation of the requirements. I'd prefer to keep it there, but we can duplicate it here as well if you think that would make it more readable.

Reply:
I don't disagree with that.

Reply:
I've updated the opt-in question to the suggested wording.